File: android/src/main/java/com/codenameakshay/async_wallpaper/VideoLiveWallpaper.java
package com.codenameakshay.async_wallpaper;

import android.annotation.SuppressLint;
import android.app.WallpaperManager;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Build;
import android.service.wallpaper.WallpaperService;
import android.text.TextUtils;
import android.util.Log;
import android.view.SurfaceHolder;

import androidx.annotation.RequiresApi;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.nio.channels.FileChannel;


public class VideoLiveWallpaper extends WallpaperService {

    public static final String VIDEO_PARAMS_CONTROL_ACTION = "com.codenameakshay.async_wallpaper";
    public static final String KEY_ACTION = "music";
    public static final boolean ACTION_MUSIC_UNMUTE = false;
    public static final boolean ACTION_MUSIC_MUTE = true;

    public static void setToWallPaper(Context context) {
        final Intent intent = new Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.putExtra(WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT, new ComponentName(context, VideoLiveWallpaper.class));
        context.startActivity(intent);
        try {
            WallpaperManager.getInstance(context).clear();
        } catch (IOException e) {
            Log.e("VideoLiveWallpaper", "Error clearing wallpaper", e);
        }
    }

    public static void openWallpaperChooser(Context context) {
        Intent intent = new Intent(WallpaperManager.ACTION_LIVE_WALLPAPER_CHOOSER);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
        try {
            WallpaperManager.getInstance(context).clear();
        } catch (IOException e) {
            Log.e("VideoLiveWallpaper", "Error clearing wallpaper", e);
        }
    }

    @Override
    public Engine onCreateEngine() {
        return new VideoEngine();
    }

    class VideoEngine extends Engine {
        private MediaPlayer mediaPlayer;
        private BroadcastReceiver broadcastReceiver;
        private boolean isReceiverRegistered = false;
        private WeakReference<Context> contextRef;

        @RequiresApi(api = Build.VERSION_CODES.TIRAMISU)
        @Override
        public void onCreate(SurfaceHolder surfaceHolder) {
            super.onCreate(surfaceHolder);
            contextRef = new WeakReference<>(getApplicationContext());
            registerVolumeReceiver();
        }

        private void registerVolumeReceiver() {
            if (!isReceiverRegistered) {
                IntentFilter intentFilter = new IntentFilter(VIDEO_PARAMS_CONTROL_ACTION);
                broadcastReceiver = new BroadcastReceiver() {
                    @Override
                    public void onReceive(Context context, Intent intent) {
                        if (mediaPlayer != null) {
                            boolean action = intent.getBooleanExtra(KEY_ACTION, false);
                            mediaPlayer.setVolume(action ? 0 : 1.0f, action ? 0 : 1.0f);
                        }
                    }
                };
                
                // Use the appropriate registration method based on Android version
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    registerReceiver(broadcastReceiver, intentFilter, Context.RECEIVER_NOT_EXPORTED);
                } else {
                    registerReceiver(broadcastReceiver, intentFilter);
                }
                
                isReceiverRegistered = true;
            }
        }

        private void releaseMediaPlayer() {
            if (mediaPlayer != null) {
                try {
                    if (mediaPlayer.isPlaying()) {
                        mediaPlayer.stop();
                    }
                    mediaPlayer.reset();
                    mediaPlayer.release();
                } catch (Exception e) {
                    Log.e("VideoEngine", "Error releasing MediaPlayer", e);
                } finally {
                    mediaPlayer = null;
                }
            }
        }

        @SuppressLint("SdCardPath")
        @Override
        public void onSurfaceCreated(SurfaceHolder holder) {
            super.onSurfaceCreated(holder);
            initializeMediaPlayer(holder);
        }

        private void initializeMediaPlayer(SurfaceHolder holder) {
            try {
                if (mediaPlayer == null) {
                    mediaPlayer = new MediaPlayer();
                    mediaPlayer.setSurface(holder.getSurface());
                    
                    // Use the appropriate file path based on Android version
                    File videoFile = new File(getFilesDir(), "file.mp4");
                    if (videoFile.exists()) {
                        mediaPlayer.setDataSource(videoFile.getAbsolutePath());
                    } else {
                        Log.e("VideoEngine", "Video file not found: " + videoFile.getAbsolutePath());
                        return;
                    }
                    
                    mediaPlayer.setLooping(true);
                    mediaPlayer.setVideoScalingMode(MediaPlayer.VIDEO_SCALING_MODE_SCALE_TO_FIT_WITH_CROPPING);
                    mediaPlayer.prepare();
                    mediaPlayer.start();
                    mediaPlayer.setVolume(0, 0);
                    Log.d("VideoEngine", "MediaPlayer started successfully");
                }
            } catch (IOException e) {
                Log.e("VideoEngine", "Error initializing MediaPlayer", e);
            }
        }

        @Override
        public void onVisibilityChanged(boolean visible) {
            if (mediaPlayer != null) {
                if (visible) {
                    mediaPlayer.start();
                } else {
                    mediaPlayer.pause();
                }
            }
        }

        @Override
        public void onSurfaceDestroyed(SurfaceHolder holder) {
            super.onSurfaceDestroyed(holder);
            releaseMediaPlayer();
        }

        @Override
        public void onDestroy() {
            super.onDestroy();
            releaseMediaPlayer();
            if (isReceiverRegistered && broadcastReceiver != null) {
                try {
                    unregisterReceiver(broadcastReceiver);
                    isReceiverRegistered = false;
                } catch (Exception e) {
                    Log.e("VideoEngine", "Error unregistering receiver", e);
                }
            }
        }
    }

    public void copyFile(File fromFile, File toFile) {
        FileInputStream fileInputStream = null;
        FileOutputStream fileOutputStream = null;
        FileChannel fileChannelInput = null;
        FileChannel fileChannelOutput = null;
        
        try {
            fileInputStream = new FileInputStream(fromFile);
            fileOutputStream = new FileOutputStream(toFile);
            fileChannelInput = fileInputStream.getChannel();
            fileChannelOutput = fileOutputStream.getChannel();
            
            long size = fileChannelInput.size();
            long transferred = 0;
            while (transferred < size) {
                transferred += fileChannelInput.transferTo(transferred, size - transferred, fileChannelOutput);
            }
        } catch (IOException e) {
            Log.e("AsyncWallpaperPlugin", "Error copying file", e);
            throw new RuntimeException("Failed to copy file", e);
        } finally {
            closeQuietly(fileChannelInput);
            closeQuietly(fileChannelOutput);
            closeQuietly(fileInputStream);
            closeQuietly(fileOutputStream);
        }
    }

    private void closeQuietly(AutoCloseable closeable) {
        if (closeable != null) {
            try {
                closeable.close();
            } catch (Exception e) {
                Log.e("AsyncWallpaperPlugin", "Error closing resource", e);
            }
        }
    }
}

File: android/src/main/kotlin/com/codenameakshay/async_wallpaper/VideoLiveWallpaper.kt
package com.codenameakshay.async_wallpaper

import android.app.WallpaperManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.media.MediaPlayer
import android.service.wallpaper.WallpaperService
import android.util.Log
import android.view.SurfaceHolder
import java.io.File

/** The two scaling modes supported directly by Android's MediaPlayer. */
enum class VideoWallpaperScaleMode {
  CENTER_CROP,
  FIT_CENTER,
}

/**
 * Live wallpaper backed by the active file managed by [VideoWallpaperRepository].
 * Player work is deliberately contained in the engine: surface and visibility
 * callbacks only transition the pure [VideoPlaybackStateMachine], then the
 * resulting action is applied while holding a single lifecycle lock.
 */
class VideoLiveWallpaper : WallpaperService() {
  override fun onCreateEngine(): Engine = VideoEngine()

  inner class VideoEngine : Engine() {
    private val lifecycleLock = Any()
    private val stateMachine = VideoPlaybackStateMachine()

    private var mediaPlayer: MediaPlayer? = null
    private var playerGeneration = 0L

    override fun onSurfaceCreated(holder: SurfaceHolder) {
      super.onSurfaceCreated(holder)
      synchronized(lifecycleLock) {
        if (stateMachine.onSurfaceCreated() == VideoPlaybackAction.PREPARE) {
          preparePlayerLocked(holder)
        }
      }
    }

    override fun onVisibilityChanged(visible: Boolean) {
      synchronized(lifecycleLock) {
        applyActionLocked(stateMachine.onVisibilityChanged(visible))
      }
    }

    override fun onSurfaceDestroyed(holder: SurfaceHolder) {
      synchronized(lifecycleLock) {
        applyActionLocked(stateMachine.onSurfaceDestroyed())
      }
      super.onSurfaceDestroyed(holder)
    }

    override fun onDestroy() {
      synchronized(lifecycleLock) {
        applyActionLocked(stateMachine.onEngineDestroyed())
      }
      super.onDestroy()
    }

    private fun preparePlayerLocked(holder: SurfaceHolder) {
      val file = File(filesDir, VideoWallpaperRepository.ACTIVE_VIDEO_FILE_NAME)
      if (!file.isFile || !file.canRead() || file.length() <= 0L) {
        Log.w(TAG, "No readable live wallpaper video at ${file.absolutePath}")
        stateMachine.onPlayerError()
        return
      }

      val player = try {
        MediaPlayer()
      } catch (error: Exception) {
        Log.e(TAG, "Unable to create live wallpaper player", error)
        stateMachine.onPlayerError()
        return
      }

      val generation = ++playerGeneration
      mediaPlayer = player
      try {
        player.setOnPreparedListener { preparedPlayer ->
          synchronized(lifecycleLock) {
            if (!isCurrentPlayer(preparedPlayer, generation)) {
              return@synchronized
            }
            applyActionLocked(stateMachine.onPrepared())
          }
        }
        player.setOnErrorListener { failedPlayer, what, extra ->
          synchronized(lifecycleLock) {
            if (isCurrentPlayer(failedPlayer, generation)) {
              Log.e(TAG, "Live wallpaper player error: what=$what extra=$extra")
              handlePlayerFailureLocked()
            }
          }
          true
        }
        player.setSurface(holder.surface)
        player.setDataSource(file.absolutePath)
        player.isLooping = true
        // Live wallpaper should never take over the user's audio by default.
        player.setVolume(0f, 0f)
        player.setVideoScalingMode(requestedScalingMode.mediaPlayerMode())
        player.prepareAsync()
      } catch (error: Exception) {
        Log.e(TAG, "Failed to prepare live wallpaper player", error)
        handlePlayerFailureLocked()
      }
    }

    private fun applyActionLocked(action: VideoPlaybackAction) {
      when (action) {
        VideoPlaybackAction.NONE -> Unit
        VideoPlaybackAction.PREPARE -> Unit // Only onSurfaceCreated owns the SurfaceHolder.
        VideoPlaybackAction.START -> startPlayerLocked()
        VideoPlaybackAction.PAUSE -> pausePlayerLocked()
        VideoPlaybackAction.RELEASE -> releasePlayerLocked()
      }
    }

    private fun startPlayerLocked() {
      val player = mediaPlayer ?: return
      try {
        player.start()
      } catch (error: Exception) {
        Log.e(TAG, "Failed to start live wallpaper player", error)
        handlePlayerFailureLocked()
      }
    }

    private fun pausePlayerLocked() {
      val player = mediaPlayer ?: return
      try {
        player.pause()
      } catch (error: Exception) {
        Log.e(TAG, "Failed to pause live wallpaper player", error)
        handlePlayerFailureLocked()
      }
    }

    private fun handlePlayerFailureLocked() {
      stateMachine.onPlayerError()
      releasePlayerLocked()
    }

    /** The only release path. Clearing the reference first makes it idempotent. */
    private fun releasePlayerLocked() {
      val player = mediaPlayer ?: return
      mediaPlayer = null
      playerGeneration += 1L
      runCatching { player.release() }
        .onFailure { error -> Log.w(TAG, "Failed to release live wallpaper player", error) }
    }

    private fun isCurrentPlayer(candidate: MediaPlayer, generation: Long): Boolean {
      return mediaPlayer === candidate && playerGeneration == generation
    }
  }

  companion object {
    private const val TAG = "VideoLiveWallpaper"

    @Volatile
    private var requestedScalingMode = VideoWallpaperScaleMode.CENTER_CROP

    /**
     * Hook for the request layer. Unsupported MediaPlayer modes intentionally
     * retain the reliable center-crop default rather than pretending to apply
     * a transform that the platform player cannot perform.
     */
    fun configureScaleMode(scaleMode: WallpaperScaleModeData?) {
      requestedScalingMode = when (scaleMode) {
        WallpaperScaleModeData.FIT_CENTER -> VideoWallpaperScaleMode.FIT_CENTER
        else -> VideoWallpaperScaleMode.CENTER_CROP
      }
    }

    fun configureScaleMode(scaleMode: VideoWallpaperScaleMode) {
      requestedScalingMode = scaleMode
    }

    fun setToWallpaper(context: Context) {
      val intent = Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        putExtra(
          WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT,
          ComponentName(context, VideoLiveWallpaper::class.java),
        )
      }
      context.startActivity(intent)
    }

    fun openWallpaperChooser(context: Context) {
      val intent = Intent(WallpaperManager.ACTION_LIVE_WALLPAPER_CHOOSER).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }
      context.startActivity(intent)
    }
  }
}

private fun VideoWallpaperScaleMode.mediaPlayerMode(): Int {
  return when (this) {
    VideoWallpaperScaleMode.CENTER_CROP -> {
      MediaPlayer.VIDEO_SCALING_MODE_SCALE_TO_FIT_WITH_CROPPING
    }
    VideoWallpaperScaleMode.FIT_CENTER -> MediaPlayer.VIDEO_SCALING_MODE_SCALE_TO_FIT
  }
}

/** Pure lifecycle policy exercised by [VideoStateMachineTest] on the JVM. */
enum class VideoPlaybackState {
  IDLE,
  PREPARING,
  PAUSED,
  PLAYING,
  RELEASED,
}

enum class VideoPlaybackAction {
  NONE,
  PREPARE,
  START,
  PAUSE,
  RELEASE,
}

class VideoPlaybackStateMachine {
  var state: VideoPlaybackState = VideoPlaybackState.IDLE
    private set

  var hasSurface: Boolean = false
    private set

  var isVisible: Boolean = false
    private set

  fun onSurfaceCreated(): VideoPlaybackAction {
    if (state == VideoPlaybackState.RELEASED) {
      return VideoPlaybackAction.NONE
    }
    hasSurface = true
    if (state != VideoPlaybackState.IDLE) {
      return VideoPlaybackAction.NONE
    }
    state = VideoPlaybackState.PREPARING
    return VideoPlaybackAction.PREPARE
  }

  fun onVisibilityChanged(visible: Boolean): VideoPlaybackAction {
    if (state == VideoPlaybackState.RELEASED) {
      return VideoPlaybackAction.NONE
    }
    isVisible = visible
    return when (state) {
      VideoPlaybackState.PAUSED if visible -> {
        state = VideoPlaybackState.PLAYING
        VideoPlaybackAction.START
      }
      VideoPlaybackState.PLAYING if !visible -> {
        state = VideoPlaybackState.PAUSED
        VideoPlaybackAction.PAUSE
      }
      else -> VideoPlaybackAction.NONE
    }
  }

  fun onPrepared(): VideoPlaybackAction {
    if (state == VideoPlaybackState.RELEASED) {
      return VideoPlaybackAction.NONE
    }
    if (!hasSurface) {
      state = VideoPlaybackState.IDLE
      return VideoPlaybackAction.RELEASE
    }
    if (state != VideoPlaybackState.PREPARING) {
      return VideoPlaybackAction.NONE
    }
    return if (isVisible) {
      state = VideoPlaybackState.PLAYING
      VideoPlaybackAction.START
    } else {
      state = VideoPlaybackState.PAUSED
      VideoPlaybackAction.NONE
    }
  }

  fun onPlayerError(): VideoPlaybackAction {
    if (state == VideoPlaybackState.RELEASED) {
      return VideoPlaybackAction.NONE
    }
    val hadPlayer = state != VideoPlaybackState.IDLE
    state = VideoPlaybackState.IDLE
    return if (hadPlayer) VideoPlaybackAction.RELEASE else VideoPlaybackAction.NONE
  }

  fun onSurfaceDestroyed(): VideoPlaybackAction {
    hasSurface = false
    if (state == VideoPlaybackState.RELEASED || state == VideoPlaybackState.IDLE) {
      return VideoPlaybackAction.NONE
    }
    state = VideoPlaybackState.IDLE
    return VideoPlaybackAction.RELEASE
  }

  fun onEngineDestroyed(): VideoPlaybackAction {
    if (state == VideoPlaybackState.RELEASED) {
      return VideoPlaybackAction.NONE
    }
    val hadPlayer = state != VideoPlaybackState.IDLE
    hasSurface = false
    state = VideoPlaybackState.RELEASED
    return if (hadPlayer) VideoPlaybackAction.RELEASE else VideoPlaybackAction.NONE
  }
}
